fixed distance reading

JojoS62 opened the pull request:
the P_013_sensordef object was created on the stack and the destructor was called after leaving the PLUGIN_INIT. That deleted the included NewPingESP8266 object and using this object failed.

added also a log output for verifying contructor/destructor calls (maybe changed to LOG_LEVEL_DEBUG?)

JojoS62 committed:
fixed distance reading

the P_013_sensordef object was created on the stack and the destructor was called after leaving the PLUGIN_INIT. That deleted the included NewPingESP8266 object and using this object failed.

TD-er commented:
At first glance, I am missing a delete of the sensordefs.
Maybe it can also be added to the map using emplace?
Then an instance will be created in the map.

JojoS62 commented:
the PLUGIN_EXIT calls an erase and that will delete the sensordef. I get the log output for deleting the sensordef when I remove the sensor from the takslist. The constructor is also called only once on creating.

JojoS62 commented:
it needs further testing, now I got a crash on deleting the sensor from the tasklist.
```
68669 : P013 exit
68669 : delete P_013_sensordef
68721 : FS   : Error while reading/writing config.dat in 1179
68748 : FILE : Saved config.dat
```

JojoS62 commented:
it looks like my config is corrupt, even after a firmware reset and starting a new configuration it get an exception after connectiing to my router. How can I clear all or is the flash broken?

TD-er commented:
There is an empty flash included.
You can flash that and then start over as a new node.

JojoS62 commented:
sorry, I dont't see that. What is the name for this file?

Grovkillen commented:
In the release zip file.

JojoS62 commented:
thanks, I'm using a new device and will check this later. The problem with the corrupted config may be caused by the bad code before.
Now it works, but the logic with adding/removing is still not ok. When I press 'submit' on the config page for the distance sensor, the constructor gets called every time. So a check if there is already an entry for a device in the map is necessary before adding a new one.

TD-er commented:
Every update of the plugin/device will also call the init.
So either you will have to add an update function to the ping class (when a pin has changed) or delete/new the object.

JojoS62 committed:
on INIT, delete sensordef before adding new

JojoS62 commented:
still difficult. Several times deleting and creating a new sensor was ok. Then I got a 'max Daily flash cycles exceeded' error and I rebooted the device. The sensor was in the config but with empty settings. After deleting this sensor, my device is corrupt now again. It looks like it is in a state where after loading the sensor is in a bad state.
Q: what happens on ESP reboot? Is the PLUGIN_INIT event called for every task in the config?

TD-er commented:
Yes, init is called for enabled tasks.
Have you tried using emplace in the map? Then you supply the parameters needed for the template object and the object will be created in place in the container.
I guess you don't have to allocate it on the heap then, and also don't need to copy it. So the destructor of the sonar pointer is not called then after the copy.

JojoS62 committed:
removed sensordefs helper structure

to simplify the code

JojoS62 commented:
I'm not sure what you mean, must check 'emplace'.
Now I have removed the sensordefs struct and I store a pointer to the NewPing object in the map. This avoids also the copy problem, but for a pointer delete is not called automatically. Maybe not the best solution, but this is working now without crashes. For testing, I had added  logging outputs to the constructor/destructor in the NewPing class.
Another option will be a smart pointer, I would prefer that solution.

TD-er commented:
Can you test this PR?
https://github.com/letscontrolit/ESPEasy/pull/1815

JojoS62 committed:
changed implementation to use shared_ptr instead of helper structure

this makes it easier to use this as a template for other sensors. No additional helper class/structure is needed.

TD-er commented:
I think the last commit is a nice fix.
Did you check it to be stable?

JojoS62 commented:
I have deleted and created the sonar several times, added two instances and rebooted the ESP several times.  Tested also the calls to sonar constructor/destructor. That worked fine without crashes.

TD-er commented:
And also multiple instances of the sensor now work? (tested with real hw I mean)

JojoS62 commented:
I have checked if the pulses were available at the ESP with a scope. For real HW, I have to prepare a levelshifter first. The obvious problem was the missing trigger pulses, these are ok now. When there are still other problems in the NewPingESP8266 class than I have not tested it yet. I can prepare the HC+level shifter until tomorrow.

JojoS62 commented:
thanks. I have tested now with 1 real sensor HC-SR04 in value and state mode and it works fine.

TD-er commented:
Great!
So this bug is finally fixed AND we have a good idea on keeping structs for multiple instances of the same plugins.
Good work!

JojoS62 commented:
yes, thanks. That was also my main interest, to fresh up my C++ knowledge. Too much .NET or other high level stuff is killing the good things :-)

TD-er commented:
Maybe we can do something similar to this plugin: https://github.com/letscontrolit/ESPEasy/issues/1486

JojoS62 commented:
Ok, will check tomorrow, now the garden is calling.
